src/fft_cufft.f90

```diff
@@ -373,21 +401,44 @@ end subroutine init_fft_engine
 !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
    !  This routine performs one-time finalisations for the FFT engine
 !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
-   subroutine finalize_fft_engine
+   subroutine finalize_fft_engine(engine)
 
       implicit none
 
+      type(decomp_2d_fft_engine), optional :: engine
+
       integer :: i, j, istat
 
-      do j = 1, 3
-         do i = -1, 2
-            istat = cufftDestroy(plan(i, j))
-            if (istat /= 0) call decomp_2d_abort(__FILE__, __LINE__, istat, "cufftDestroy")
+      if (present(engine)) then
+
+         do j = 1, 3
+            do i = -1, 2
+               istat = cufftDestroy(engine%plan(i, j))
+               if (istat /= 0) call decomp_2d_abort(__FILE__, __LINE__, istat, "cufftDestroy")
+            end do
          end do
-      end do
+
+      else
+
+         nullify (plan)
+         nullify (cufft_workspace)
+
+      end if
```

Comment:
Should there be a check on if `ngrid > 1` in which case this should loop over all engines (I think)?

Reply:
All the engines are cleaned before cleaning the variables `plan` and `cufft_workspace`. This is performed in `decomp_2d_fft_finalize`, see fft_common.f90  : 

```
subroutine decomp_2d_fft_finalize

   implicit none

   integer :: igrid

   ! Clean each engine
   do igrid = 1, n_grid
      call fft_engines(igrid)%fin()
   end do

   ! Free memory
   n_grid = 0
   deallocate (fft_engines)

   ! Nullify all the pointers in the module
   nullify (format)
   nullify (nx_fft)
   nullify (ny_fft)
   nullify (nz_fft)
   nullify (ph)
   nullify (sp)
   nullify (wk2_c2c)
   nullify (wk2_r2c)
   nullify (wk13)
   nullify (inplace)
   nullify (inplace_r2c)
   nullify (inplace_c2r)

   ! Clean engine-specific stuff located in the module
   call finalize_fft_engine()

end subroutine decomp_2d_fft_finalize
```